### DISCUSSION TITLE
[BE] Exception 네이밍, 표준 예외 vs 커스텀 예외

### DISCUSSION THREAD

**Discussion by yenawee** [woowacourse-teams/2023-stamp-crush#212] [General]
현재 NotFoundException 등 우리가 만들어둔 커스텀 예외들이 현재 스프링 JPA 의 예외와 네이밍이 겹쳐서 패키지를 잘 확인하고 넣어야 하는 불편함이 있습니다 
네이밍을 바꿔야 합니다

좋은 의견 구합니다

**Comment by gitchannn** [comment]
### 이 상태로는 안되고, 바꿔야 한다는 것에는 동의함.

**Reply by gitchannn** [reply]
1. 모든 커스텀 예외 앞에 StampCrush를 붙인다.

**Comment by youngh0** [comment]
JPA에서 제공하는 NotFoundException을 CafeNotFoundException이런 애들이 상속받게 하면 예외핸들러에서 NotFoundException하나만 잡아서 처리할 수 있지 않을까 라고 생각합니다.

**Comment by jundonghyuk** [comment]
커스텀예외를 써야할까 ? 
이미 제공해주는 예외가 많은데, 그냥 써도되지 않을까 ? maybe..?

**Reply by gitchannn** [reply]
표준예외파시군요~

**Comment by yenawee** [comment]
우리가 만들어둔 최상위 예외 StampCrushException 을 따로 활용하고 있지 않은거같아서
표준 예외를 사용하는 것이 나아보입니다.

**Comment by gitchannn** [comment]
# 각자 의견을 써주세요

**Reply by gitchannn** [reply]
**깃짱의 의견**
- 현재 우리팀은 프론트엔드와 에러 코드를 주고받지는 않는 상황
    - API 설계 당시부터 에러 코드를 고려하지는 않았으므로, 지금에서야 **에러 코드를 도입하는 것은 시간적으로 무리가 있고 큰 효과를 일으키기 어려울 것 같다**고 생각함. 
- 우리팀은 로그를 많이 찍어두지는 않았지만, 에러 메세지에 넣은 내용을 그대로 로그로 사용하기로 결정함.
    - 로그를 사용하는 경우에는 표준 예외든 커스텀 예외든 큰 상관 없음. 하지만!!! (아래)
- 프론트엔드 측에 에러 메세지를 넘겨줄 지에 대해서
    - 하디는 프론트엔드에서 메세지를 사용하지 않아서 큰 의미가 없다고 했지만, 반드시 코드에서 사용하는 것이 아니라 API를 호출하고 나서 발생하는 문제에 대해서 백엔드와 소통만으로도 충분히 의미가 있다고 생각함. => **_메세지는 내려주자_**
    - 표준 예외에 내 메세지를 담아서, ExceptionHandler에서 일괄 처리하는 경우에, 내가 작성한 메세지가 아닌 애플리케이션에서 그대로 넘어온 것에 _민감 정보가 담긴 상태로 클라이언트에 넘어갈 수 있으므로_, **커스텀 예외를 그대로 사용해야 한다고 생각함**

결론: `커스텀 예외 쓰고, 커스텀 예외 내용을 로그로도 찍고 프론트엔드에도 보내주자! 표준 예외의 메세지는 프론트엔드에 숨기자!`

**Reply by jundonghyuk** [reply]
현재 프론트에서 예외 분기처리 안하고 있음
-> 현재는 메시지로 내려줘도 되는 상황 

만약 프론트에서 예외 분기처리를 하고 싶다는 요구 사항이 들어오면
-> 문자열보다는 약속된 숫자로 맞추는게 나을 것 같음.

결론: `현재의 기조를 유지하고(커스텀 예외를 쓰고) 메시지를 구체화하자(로그찍기위해) `

**Reply by yenawee** [reply]
지금 고쳐야 할 사항도 많고 이미 커스텀 예외를 거의 모든 곳에서 쓰고 있으니 지금 방향대로 가되

클라이언트에 에러메세지는 이번에 개발하면서 response 를 많이 확인해서 에러를 잡았기 때문에 보내주는 것이 좋다고 생각함 
(로그 찍히는 메세지처럼 자세할 필요는 없고 커스텀해서)

그리고 로그를 빨리 찍어둬야 한다는 생각이 들었습니다

**Reply by youngh0** [reply]
### 커스텀 예외 찬성
**이유**
- stamp찍을 수 있는 최대 개수를 초과해서 들어오는 입력의 경우 같이 표준 예외로 처리하기에는 다소 명확하지 못한 부분이 있는 것 같다.
- 현재 exceptionHandler에서 exception.getMessage()를 넘겨주는데 표준예외만 사용할 경우 spring이 표준예외를 던졌을 때 노출되선 안되는 정보가 노출될 가능성이 있다.
    - Spring이 던진 표준예외는 로그만 찍고, 우리가 던진 커스텀 예외 메시지는 로그도 찍고 프론트한테도 반환해주고 이런 방식

### 커스텀 예외 사용 방향
- 현재 너무 세세하게 커스텀 예외가 존재하는 것 같다.
- BusinnessException, NotFoundException 정도의 예외를 만들고 예외메시지를 자세하게 작성하면 커스텀 예외 관리 측면에서 조금 더 수월할 것 같다.

**Comment by gitchannn** [comment]
# 결론

### 커스텀 예외 쓴다.
### 커스텀 예외, 표준 예외 둘 다 그대로 로그 찍는다.
### 커스텀 예외 메세지는 프론트엔드에게 넘겨준다.
### 표준 예외 메세지는 프론트엔드에 숨긴다. `eg. 알 수 없는 에러가 발생했습니다. 잠시 후 다시 시도해주세요`
### 커스텀 예외 메세지 진짜 자세하게 쓴다. `eg. Customer id: 1, 비밀번호가 맞지 않아 실패`

**Reply by yenawee** [reply]
로그는 핸들러에서 찍는다
